blogs/2021/09/06/bracket-pair-colorization.md

```diff
@@ -0,0 +1,311 @@
+Note that this approach is significantly different from data structures that are directly indexed by lines (such as using a string array to describe the line contents of a document). In particular, this approach allows to do a single binary search across and within lines.
+
+Adding two such lengths is easy, but requires a case distinction: While the line counts are added directly, the column count of the first length is only included if the second length spans zero lines.
+
+Surprisingly, most of the code does not need to be aware of how lengths are represented. Only the position mapper got significantly more complex, since care had to be taken that a single line can contain multiple text edits.
+
+As an implementation detail, we encode such lengths in a single number to reduce memory pressure. JavaScript supports integers up to $2^{53} - 1$, so we can use up to 26 bits each for the number of lines and columns.
+Unfortunately, v8 stores numbers larger than $2^{31}$ [on the heap](https://v8.dev/blog/react-cliff#smi-heapnumber-mutableheapnumber), so this encoding trick did not turn out as effective as we thought.
+
+## The Devil Is In The Details: Unclosed Bracket Pairs
+
+So far, we assumed that all bracket pairs are balanced. However, we also want to support unclosed and unopened bracket pairs.
+The beauty of a recursive-decent parser is that we can use anchor sets to improve error recovery.
+
+Consider the following example:
+```
+( @1
+} @2
+) @3
+```
+
+Clearly `}` at @2 does not close any bracket pair and represents an unopened bracket. The brackets at @1 and @3 match nicely.
+However, when inserting `{` at the beginning of the document, the situation changes:
+```
+{ @0
+( @1
+} @2
+) @3
+```
+Now, @0 and @2 should be matched, while @1 is an unclosed bracket and @3 an unopened bracket.
+
+In particular, @1 should be an unclosed bracket terminating before @2 in the following example:
+```
+{
+    ( @1
+} @2
+{}
+```
+
+Otherwise, opening a parenthesis could change the nesting-level of unrelated following bracket pairs.
+
+To support this kind of error recovery, anchor sets can be used to track the set of expected tokens the caller can continue with. At position @1 in the previous example, the anchor set would be $\{$ `}` $\}$. Thus, when parsing the bracket pair at @1 finds the unexpected bracket `}` at @2, it does not consume it and returns an unclosed bracket pair.
+In the very first example, the anchor set at @2 is $\{$ `)` $\}$, but the unexpected character is `}`. Because it is not part of the anchor set, it is reported as unopened bracked.
+
+This needs to be considered when reusing nodes: The pair `( } )` cannot be reused when prepending it with `{`! We use bit-sets to encode anchor sets and compute the set of containing unopened brackets for every node. If they intersect, we cannot reuse the node. Luckily, there are only a few bracket types, so this does not affect performance too much.
+
+## Outlook
```

Comment:
Wait, bar charts please! I suggest to put at least as much effort as in the big O analysis into measuring the times it takes to colorize or the used up memory in `checker.ts`, `sqlite3.c`, maybe a couple more like `bootstrap.css`. I would also compare with 2-3 extensions, like bracketpair1, bracketpair2, rainbow brackets. You can also compare it with the previous bracket matching implementation that we still have.